src/solver/gemm_common.cpp

```diff
@@ -0,0 +1,112 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/env.hpp>
+#include <miopen/solver/gemm_common.hpp>
+
+#include <algorithm>
+#include <limits>
+
+// Vega20 fails to access GPU memory larger than the
+// return value of GetMaxMemoryAllocSize() of Vega10.
+#define WORKAROUND_MLOPEN_ISSUE_1430 1
+
+MIOPEN_DECLARE_ENV_VAR_BOOL(MIOPEN_WORKAROUND_ISSUE_2808)
+MIOPEN_DECLARE_ENV_VAR_BOOL(MIOPEN_WORKAROUND_ISSUE_2809)
+
+namespace miopen {
+namespace solver {
+namespace conv {
+namespace gemm {
+
+std::size_t MaxMemAllocSz(Handle& h, const miopen::conv::ProblemDescription& problem)
+{
+    constexpr auto wa_2808_16bit_limit    = static_cast<std::size_t>(7287183769);
+    constexpr auto wa_2808_other_limit    = wa_2808_16bit_limit * 2;
+    constexpr auto wa_2809_fp32_wrw_limit = wa_2808_16bit_limit;
+
+    const auto m = h.GetMaxMemoryAllocSize();
+    auto limit   = std::numeric_limits<std::size_t>::max();
+
+#if WORKAROUND_MLOPEN_ISSUE_1430
+    const auto device = h.GetTargetProperties().Name();
+    if(device == "gfx906" || device == "gfx900")
+    {
+        // Workaround for MLOpen issue 1430. Vega20 fails to access GPU memory
+        // larger than the return value of GetMaxMemoryAllocSize() of Vega10.
+        limit = std::min(limit, static_cast<std::size_t>(7287183769));
```

Comment:
```suggestion
        limit = std::min(limit, wa_2808_16bit_limit);
```

Reply:
I want independent value here.

Reply:
Then declare it properly)
```
if (42 == status)
```

Reply:
There is comment that explains the semantics of the literal, and the literal is used only once.

BTW 42 has well-known meaning ))

Reply:
@CAHEK7 BTW are you insisting or just discussing?

Reply:
I just want to have everything consistent.
All those values are used only once and instead of declaring them, `wa_2808_16bit_limit` can be used directly.
```
    constexpr auto wa_2808_other_limit    = wa_2808_16bit_limit * 2;
    constexpr auto wa_2809_fp32_wrw_limit = wa_2808_16bit_limit;
```
But this literal looks exactly the same as `wa_2808_16bit_limit` and it look suspicions even if it has a comment, because the comment does not clarify anything - is `7287183769` the limit of `GetMaxMemoryAllocSize()`? Why it is hardcoded if there is a function which returns that value? 
Why it says Vega20, then refers to Vega10?

Reply:
@CAHEK7 
> But this literal looks exactly the same as wa_2808_16bit_limit...

I am not changing existing limits because 
- We do not know the root reason of the 2808 and 2809
- Increasing the value may lead to correctness errors
- Decreasing the value may lead to performance issues

That's why only limit for FP32 Fwd/Bwd was changed (and carefully tested).

The values for targets other than Vega10/20 will be changed (and hopefully removed at all) after the full-blown fix.

> Why it is hardcoded if there is a function which returns that value?
> Why it says Vega20, then refers to Vega10?

As comment explains, we found that "Vega20 fails to access GPU memory larger than the return value of GetMaxMemoryAllocSize() of Vega10". `7287183769` is the value that GetMaxMemoryAllocSize() returns on Vega10.

Reply:
I'm not asking about changing the numbers, I'm talking about magic numbers which are not clear from the code. And I'm talking about the fact that you named some of them and did not some others.
That explanation is great and useful, but it will stay here in the comments to the pull request, not in the code.

It's up to you to name it properly or leave everything as it is - I approved this PR since the beginning.

Reply:
@CAHEK7 it's your right (and even duty) to ask questions if something is unclear.

> And I'm talking about the fact that you named some of them and did not some others.

The reasons for not naming literal under WORKAROUND_MLOPEN_ISSUE_1430 are:
- It's hard to invent name that explains the semantics better than the existing comment
- the literal is used only once
- the w/a it is not intended to be fixed and will be removed together with gfx906 support.

Thanks for reviewing!